klam-ssh support for CoreOS

eadasiak opened the pull request:
This adds support for klam-ssh on CoreOS.

The klam-ssh.sh script in v3/setup installs the pre-reqs:
1.  A replacement for /etc/nsswitch.conf that includes the libnss_ato library method for passwd, shadow, and group
2.  The klam-ssh.conf file in /opt/klam/lib
3.  The libnss_ato shared library in /opt/klam/lib64, and the ld.so.conf.d config for that file path
4.  Reconfigures the linker cache
5.  Adds the klamfed user (all klam-ssh users are squashed to this user)
6.  Adds sudo support for klamfed
7.  Moves supporting scripts into place in /opt/klam/lib
8.  Appends config to the sshd_config file for the AuthorizedKeysCommand

The authorizedkeys_command.sh script runs the klam-ssh docker container to get the user's ssh key from S3

The downloadS3.sh script runs the klam-ssh docker container to get the list of authorized klam-ssh users via S3.

Let me know if you have any questions.  Next stop -- the bastion host!

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Initial check-in of klam-ssh support

Commit:
klam-ssh additions

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Last few refinements

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
make scripts executable

bryanlatten commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,137 @@
+
+if [[ $REGION == "eu-west-1" ]]; then
+    KEY_LOCATION="-ew1"
+elif [[ $REGION == "ap-northeast-1" ]]; then
```
should we make this list accessible to the widest swath of regions, or is unnecessary?

eadasiak replied:
Unfortunately, I was limited to the regions where klam-ssh is configured.  Version 2.0 of klam-ssh will be more widely available, and will get addressed with its release.

bryanlatten commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,137 @@
+#!/bin/bash
```
should we slap an `-e` on there to exit if any command fails?

chr0n1x replied:
agreed. I'm also starting to lean towards `-xe` due to the amount of stuff going on in here

eadasiak replied:
Good point!  Made that change.

chr0n1x commented on v3/profile.d/klam.sh:
```diff
@@ -0,0 +1,2 @@
+KLAM_USER=$(who -m | awk '{print $1}')
+PROMPT_COMMAND='RETRN_VAL=$?;logger -p local6.debug "$KLAM_USER [$$]: $(history 1 | sed "s/^[ ]*[0-9]\+[ ]*//") [$RETRN_VAL]"'
```
this is some...voodoo magic.

eadasiak replied:
Got rid of this and made a change to the global command prompt instead to use the `who -m` voodoo instead.  This wasn't working correctly under CoreOS as it was, so away it goes!

chr0n1x commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,137 @@
+AZ=$(curl -s http://169.254.169.254/latest/meta-data/placement/availability-zone)
+REGION=${AZ::-1}
+ROLE_NAME="$(etcdctl get /klam-ssh/ROLE_NAME)"
+ENCRYPTION_ID="$(etcdctl get /klam-ssh/ENCRYPTION_ID)"
```
lowercase & dashes for all of these

matthewdfuller replied:
Also, put these in `/klam-ssh/config/your-value-here`

eadasiak replied:
OK, I fixed these up.  There will need to be an update to infrastructure as well so that the example configs have the values corrected.

chr0n1x commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,137 @@
+    exit 1
+fi
+
+# create nsswitch.conf
```
this is a huge nitpick but there _has_ to be a better way to do this

eadasiak replied:
If "unhappy"; then
  how about a case statement?
elif "unsatisfied"; then
  how about a case statement?
else
  you get a case statement
  exit 0

matthewdfuller commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,137 @@
+echo "netgroup:   nisplus" >> /home/core/nsswitch.conf
+echo "bootparams: files" >> /home/core/nsswitch.conf
+echo "automount:  files nisplus" >> /home/core/nsswitch.conf
+echo "aliases:    files nisplus" >> /home/core/nsswitch.conf
```
Another stylistic nitpick, but can't we do something like:

```
cat <<EOT >> /home/core/nsswitch.conf
passwd: files usrfiles ato
shadow: files usrfiles ato
...
EOT
```

Just makes things easier to read / cleaner.

chr0n1x replied:
👍

eadasiak replied:
You got it!

matthewdfuller commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,137 @@
+echo "    \"resource_location\": \"amazon\"," >> /home/core/klam-ssh.conf
+echo "    \"time_skew\": \"permissive\"," >> /home/core/klam-ssh.conf
+echo "    \"s3_region\": \"${REGION}\"" >> /home/core/klam-ssh.conf
+echo "}" >> /home/core/klam-ssh.conf
```
Same here: `cat <<EOT >> /home/core/klam-ssh.conf`

matthewdfuller commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,137 @@
+mv /home/core/mesos-systemd/v3/util/downloadS3.sh /opt/klam/lib
+
+# Restart SSHD
+systemctl restart sshd.service
```
Is this going to conflict with the existing service which restarts sshd? Can we reduce how many times we are restarting it by making sure all the conf changes are made first, then restarting it once?

chr0n1x replied:
I don't _think_ so? It shouldn't just because this will run serially. @eadasiak can you test and confirm?

matthewdfuller replied:
Conflict was probably the wrong word; I'm thinking more of the fact that ssh is being restarted twice. We already have to wait about five minutes before we can SSH in to an instance, this might make that longer.

chr0n1x replied:
The restart is actually quite fast though. We see the daemon restart when fleet is restarted and it's near instantaneous to establish a new connection

eadasiak replied:
I wasn't able to find where the other sshd restart happens.  This particular restart is really fast, but if the sequence of events is such that the setup scripts occur before the other service restart, I can strike this one.

matthewdfuller replied:
I'm okay with leaving this here; was just thinking about optimizing it, but if you say it's fast, LGTM

matthewdfuller commented on v3/util/authorizedkeys_command.sh:
```diff
@@ -0,0 +1,12 @@
+ENCRYPTION_ID=$(etcdctl get /klam-ssh/ENCRYPTION_ID)
+ENCRYPTION_KEY=$(etcdctl get /klam-ssh/ENCRYPTION_KEY)
+ROLE_NAME=$(etcdctl get /klam-ssh/ROLE_NAME)
+KEY_LOCATION_PREFIX=$(etcdctl get /klam-ssh/KEY_LOCATION_PREFIX)
```
Make sure to update these to `/klam-ssh/config/your-value-here` as well.

eadasiak replied:
Done!

eadasiak replied:
Associated infrastructure PR https://github.com/adobe-platform/infrastructure/pull/71

matthewdfuller commented on v3/util/authorizedkeys_command.sh:
```diff
@@ -0,0 +1,12 @@
+ENCRYPTION_KEY=$(etcdctl get /klam-ssh/ENCRYPTION_KEY)
+ROLE_NAME=$(etcdctl get /klam-ssh/ROLE_NAME)
+KEY_LOCATION_PREFIX=$(etcdctl get /klam-ssh/KEY_LOCATION_PREFIX)
+IMAGE=$(etcdctl get /images/klam-ssh)
```
`'images/klam-ssh` should be set (as a default) in `v3/setup/klam-ssh.sh`

eadasiak replied:
Ah, I see your point.  I was relying on the configs from the infrastructure repo to do this instead.  I've set a default value in v3/setup/klam-ssh.sh

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
update from master; fixes from comments in PR

Commit:
Modify global prompt for klam-ssh

uses who -m instead of \u in the command prompt.  Since users are
squashed to a single user (klamfed) via libnss_ato, everyone except for
core had a prompt of 'klamfed@…'

killed off klam.sh since it doesn’t really work in CoreOS

Commit:
setting default klam-ssh image value

matthewdfuller commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,144 @@
+bootparams: files
+automount:  files nisplus
+aliases:    files nisplus
+EOT
```
So much cleaner! :)

matthewdfuller commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,144 @@
+
+AZ=$(curl -s http://169.254.169.254/latest/meta-data/placement/availability-zone)
+REGION=${AZ::-1}
+ROLE_NAME="$(etcdctl get /klam-ssh/config/role-name)"
```
If the KLAM Role is created in the infrastructure stack, it should be added as an environment variable here: https://github.com/adobe-platform/infrastructure/blob/master/v2/src/userdata/templates/write_files/environment/shared.yml

Otherwise, we will have to have users pull their KLAM role out of the infrastructure stack and insert it into secrets.

chr0n1x commented on v3/setup/klam-ssh.sh:
```diff
@@ -0,0 +1,144 @@
+cat /etc/ssh/sshd_config
+
+# Change ownership of authorizedkeys_command
+chown root:root /home/core/mesos-systemd/v3/util/authorizedkeys_command.sh
```
huh...does this actually work? And why is this needed?

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
etcd2 setup for bastion host as well

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Bastion support in mesos-systemd

Commit:
Fixing klam-ssh installation issues due to reliance on values from seed-etcd.sh

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
something's not right...

Commit:
Revert "something's not right..."

This reverts commit 0d8aa6fcf848b99d3be5fccd11d43db831267df7.

Commit:
Rearranging the script order

Commit:
update-ssh-keys is causing problems

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Bastion host ssh helper script

This will generate the correct ssh config file for a user’s workstation
to connect through the bastion to the internal hosts

Commit:
Merge remote-tracking branch 'adobe-platform/master'

Commit:
Updating the ssh client config script to use the Bastion as a forward agent
